File: leet_code/problems/easy/python/014.py
"""
Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".
```
Example 1:
Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:
Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
```
"""

list1 = ["flower","flow","flight"];
list2 = ["dog","racecar","car"]

def longestCommonPrefix_1(strs):
    if not strs:
        return ""
    prefix = ""
    for i in range(len(strs[0])):
        char = strs[0][i]
        for s in strs[1:]:
            if i >= len(s) or s[i] != char:
                return prefix
        prefix += char
    return prefix



def longestCommonPrefix_2(strs):
    if not strs:
        return ""
    prefix = strs[0]
    for s in strs[1:]:
        while not s.startswith(prefix):
            prefix = prefix[:-1]
            if not prefix:
                return ""
    return prefix



print(longestCommonPrefix_1(list1));
print(longestCommonPrefix_1(list2));

print(longestCommonPrefix_2(list1));
print(longestCommonPrefix_2(list2));

File: leet_code/problems/easy/javascript/014.js
/*
Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".
```
Example 1:
Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:
Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
```
*/

const arr1 = ["flower","flow","flight"];
const arr2 = ["dog","racecar","car"];

function longestCommonPrefix_1(strs) {
    if (strs.length === 0) return "";
    let prefix = "";
    for (let i = 0; i < strs[0].length; i++) {
        const char = strs[0][i];
        for (let j = 1; j < strs.length; j++) {
            if (i >= strs[j].length || strs[j][i] !== char) {
                return prefix;
            }
        }
        prefix += char;
    }
    return prefix;
}


const longestCommonPrefix_2 = strs => {
    if (!strs.length) return "";
    return strs.reduce((prefix, str) => {
        while (str.indexOf(prefix) !== 0) {
            prefix = prefix.slice(0, -1);
            if (!prefix) return "";
        }
        return prefix;
    }, strs[0]);
};


console.log(longestCommonPrefix_1(arr1));
console.log(longestCommonPrefix_1(arr2));

console.log(longestCommonPrefix_2(arr1));
console.log(longestCommonPrefix_2(arr2));